contracts/libraries/AddressPermissionControl.sol

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: BSL-1.1
+pragma solidity 0.8.9;
+
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import "../libraries/ExceptionsLibrary.sol";
+
+contract AddressPermissionControl {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    uint256 internal _stagedToCommitAt;
+
+    mapping(address => uint256) private _stagedPermissionMasks;
+    mapping(address => uint256) private _permissionMasks;
+    EnumerableSet.AddressSet private _stagedAddresses;
+    EnumerableSet.AddressSet private _addresses;
+
+    event StagedGrantPermissions(address indexed sender, address indexed target, uint8[] permissionIds, uint256 delay);
+    event RevokedPermissionsInstant(address indexed sender, address indexed target, uint8[] permissionIds);
+    event RolledBackStagedPermissions(address indexed sender);
+    event CommittedStagedPermissions(address indexed sender);
+
+    function addresses() public view returns (address[] memory) {
+        return _addresses.values();
+    }
+
+    function addressesLength() public view returns (uint256) {
+        return _addresses.length();
+    }
+
+    function addressAt(uint256 index) public view returns (address) {
+        return _addresses.at(index);
+    }
+
+    function permissionMaskOf(address target) public view returns (uint256) {
+        return _permissionMasks[target];
+    }
+
+    function stagedAddresses() public view returns (address[] memory) {
+        return _stagedAddresses.values();
+    }
+
+    function stagedAddressesLength() public view returns (uint256) {
+        return _stagedAddresses.length();
+    }
+
+    function stagedAddressAt(uint256 index) public view returns (address) {
+        return _stagedAddresses.at(index);
+    }
+
+    function stagedPermissionMaskOf(address target) public view returns (uint256) {
+        return _stagedPermissionMasks[target];
+    }
+
+    function _hasPermission(address addr, uint8 permissionId) internal view returns (bool) {
+        return _permissionMasks[addr] & _permissionIdToMask(permissionId) != 0;
+    }
+
+    function _hasStagedPermission(address addr, uint8 permissionId) internal view returns (bool) {
+        return _stagedPermissionMasks[addr] & _permissionIdToMask(permissionId) != 0;
+    }
+
+    function _isStagedToCommit() private view returns (bool) {
+        return _stagedToCommitAt != 0;
+    }
+
+    function _permissionIdToMask(uint8 permissionId) private pure returns (uint256) {
+        return 1 << (permissionId);
+    }
+
+    function _clearStagedPermissions() private {
+        uint256 length = _stagedAddresses.length();
```

Comment:
I think `delete _stagedAddresses` should work and we won't need a separate function _clearStagedPermissions

Reply:
nope

Reply:
you can't delete a mapping this way